architecture/dns.md

```diff
@@ -19,159 +19,182 @@ We will implement a DNS module supporting the following functionality:
+    // up to the global top-level domain
+    Path() string
+
+    // Returns the subdomain under this domain as specified by the given string
+    GetSubDomain(path string) DomainEntry
+    
+    // Returns the content under this domain as specified by path String
+    // Content may either be in this domain or in a subdomain
+    // Returns latest version by default
+    GetContent(path string) ContentEntry
+
+    // Returns content at a given sequence
+    GetContentAtSequenct(path string, seq uint64) ContentEntry
+
+    // Methods to add/update/delete subdomains and content
+    AddSubDomain(domain DomainEntry) error
+    AddPrecommit(name string, precommit []byte) error
+    AddContent(c ContentEntry, reveal uint64) error
+    UpdateSubDomain(name string, domain DomainEntry) error
+    UpdateContent(newHash []byte) error
+    DeleteSubDomain(name string) error
+    DeleteContent(name string)
 }
 
-func (pe PrefixEntry) GetContentHash() []byte {
-    return nil
+// exported.go
+type DomainOwner interface {
+    // authenticates add/update/deletes to subdomains of this domain
+    AuthenticateDomainChanges(sdk.Msg) error
+    // authenticates add/update/deletes to direct content in this domain
+    AuthenticateContentChanges(sdk.Msg) error
 }
 
-// TODO: naming, InformationEntry? LatestEntry? 
-type ReverseEntry struct {
-    Owners []EntryOwner
-    Domain Domain
-    LatestSequence uint64
+// exported.go
+type ContentEntry interface {
+    Name() string
+    Path() string
+    Parent() DomainEntry
+    GetContentHashes() [][]byte
+
+    // returns content at latest Version
+    GetContent() []byte
+    GetContentAtVersion(seq uint64) []byte
 }
+```
 
-// precommit.go
-type PreCommit struct {
-    Owners []EntryOwners
-    Hash []byte   
-}
+In order to prevent frontrunning of content registration, we implement a simple commit-reveal scheme.
 
-// domain.go
-type Domain struct {
-    GetPrefix() string
-    GetContentName() string
-    GetBytes() []byte // []byte(prefix + content name)
-}
+Before adding content under a given domain and name, the domain owner must authenticate a precommit of the content, 
+which will get stored under `{PrecommitPrefix}/{Domain.Path()}/Name} -> Precommit`. Only one precommit may be stored for a 
+given content path.
+
+```go
+type Precommit struct {
+    Name string
 
-// EntryOwner is owner over a specific entry. An address may own many entries.
-type EntryOwner {
-    Address sdk.AccAddress // owner address
-    PrefixOwnershipIndex uint64 // index within the set of prefix owners
+    Precommit []byte
 }
 ```
 
 ## Msgs
 
-We will define three Msg types, `MsgRegisterPrefix`, `MsgPreCommitEntry` and `MsgCommitEntry`.
-Updates to entries will use `MsgPreCommitEntry` and `MsgCommitEntry`
+We will define three Msg interfaces, `MsgRegisterDomain`, `MsgPreCommit` and `MsgCommitContent`.
+Updates to entries will use `MsgPreCommitEntry` and `MsgCommitContent`. Deletions of content/subdomains will include 
```

Comment:
```suggestion
Updates to content will use `MsgPreCommitEntry` and `MsgCommitContent`. Deletions of content/subdomains will include 
```

we don't need to use commit-reveal for adding/updating domains, just for content